node-gyp: allow node.exe/iojs.exe to be renamed

piscisaureus opened the pull request:
On Windows, when node or io.js attempts to dynamically load a compiled
addon, the compiled addon tries to load node.exe or iojs.exe again -
depending on which import library the module used when it was linked.
This makes it impossible to rename node.exe or iojs.exe, because when
that happens the module can't find its dependencies.

With this patch, a delay-load hook is added to all modules that are
compiled with node-gyp. The delay-load hook ensures that whenever a
module tries to load imports from node.exe/iojs.exe, it'll just refer
back to the process image, thus making it possible to rename the
iojs/node binary.

Bug: (amongst others) https://github.com/iojs/io.js/issues/751
Upstream PR: https://github.com/TooTallNate/node-gyp/pull/599

R=@bnoordhuis
R=@rvagg
R=@iojs/platform-windows

am11 commented:
This is outstanding. Thank you! :+1: :star2: 

Is it possible that we could rename `iojs.exe` to anything other than `node.exe`? 
(ex. `jsl.exe (next to cl.exe)` or `jsrunner.exe`)

piscisaureus commented:
@am11 yes, you can rename it to anything you like now.

rvagg commented:
it's a bit beyond me, so windozy, but lgtm none the less

can someone verify a few addons on windows with this? I don't have time to fiddle right now unfortunately

piscisaureus commented:
Here's the ones I tested myself:
```
bson\build\Release\bson.node
node-sass\build\Release\binding.node
ws\node_modules\bufferutil\build\Release\bufferutil.node
ws\node_modules\utf-8-validate\build\Release\validation.node
```

piscisaureus commented:
@rvagg Made it a teensy bit less windozy, but unfortunately this is sort of as good as it gets...

am11 commented:
I have also tested with node-sass binary and it is working perfectly fine.

@rvagg, we are using pangyp with node-sass and so do other folks. Once this get merged here, should we apply this patch to pangyp as well?

othiym23 commented:
@piscisaureus I'm keeping an eye on @TooTallNate/node-gyp#599 now, but this is going to need to be added to the floating patch we apply to npm when it's merged to io.js until such time as a new version of `node-gyp` is released upstream, no? Can you do me a favor and ping me on issues that touch things under `deps/npm` so I know they need to be preserved on merges?

piscisaureus commented:
@othiym23 
I think we're currently floating two patches:
* this one :)
* https://github.com/iojs/io.js/commit/58a612ea9d2ede45d138174987fd8fae37f5d5d3

othiym23 commented:
OK! You're going on the reviewers list for npm updates now, because I'm not sure I'll remember this (although I am making a note of it).

Fishrock123 commented:
@othiym23 @piscisaureus though next update, with should just merge the commits again for floating, if possible.

othiym23 commented:
@Fishrock123 
> should just merge the commits again for floating, if possible.

Can you expand? I'm not sure I get what you're saying.

piscisaureus commented:
@othiym23 My guess is that we can get this patch into node-gyp easily; it benefits both node and iojs.

As for the other changes, they have been [PR-ed](https://github.com/TooTallNate/node-gyp/pull/564) to node-gyp, but the issue is that (upon insistence of you CEO - go have some fun during lunch tomorrow :smiling_imp:) iojs is currently undetectable for node-gyp, so it'd have to break node in order to support iojs.

othiym23 commented:
I understand and largely agree with @isaacs's reasoning here, and am waiting for [other changes](https://github.com/iojs/io.js/pull/493) to land so we can support this cleanly in `node-gyp` and Node.js. I just need a little help keeping the books balanced until that happens.

Fishrock123 commented:
@othiym23 sorry, wrong word. Squash. 58a612e is a combination of 3 previous commits. A new squashed floating patch would preferably be made containing this one also. (I can do that if it makes it easier for you.) :)

othiym23 commented:
I've been doing this long enough and the "floating patch" has changed enough times that I may set up a quilt. The problem with continually squashing all the patches down into one is that presumably some of them are going to be rendered round anyway some point, and then cherry-picking is going to fail. This is the exact situation quilt was made for. We can then merge the patches as a single commit, or integrate them directly into the npm update commit itself (although I think that's going a step too far – we should leave a clue in the commit log that we're not committing deps as-is.

piscisaureus commented:
I disagree with squasing. It was fair in case of 58a612e because it squashes three commits that are all half-assed solutions to making node-gyp support iojs. The above patch is only tangentially related to that.

@othiym23
An easy trick for merging is to create an "upgrade" branch at the last point were NPM was "vanilla", then upgrade NPM there, and then merge it back:

```
D:\iojs\deps>git log -n5 --oneline -- npm
3d46fef win,node-gyp: allow node.exe/iojs.exe to be renamed
269e46b deps: make node-gyp work with io.js
b542fb9 deps: upgrade npm to 2.7.3
65d0a8e deps: make node-gyp work with io.js
7d0baf1 deps: upgrade npm to 2.7.1

D:\iojs\deps>git branch upgrade-npm b542fb9

D:\iojs\deps>git checkout upgrade-npm
Switched to branch 'upgrade-npm'

D:\iojs\deps>curl -L https://github.com/npm/npm/archive/v2.7.4.tar.gz | tar xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   115    0   115    0     0    198      0 --:--:-- --:--:-- --:--:--   198
100 2601k  100 2601k    0     0   182k      0  0:00:14  0:00:14 --:--:--  392k

D:\iojs\deps>rm -rf npm

D:\iojs\deps>mv npm-2.7.4 npm

D:\iojs\deps>cd npm

D:\iojs\deps\npm>npm install
npm WARN engine follow@0.11.4: wanted: {"node":"0.10.x || 0.8.x"} (current: {"no
de":"1.6.2","npm":"2.7.1"})
-
> npm@2.7.4 prepublish D:\iojs\deps\npm
> node bin/npm-cli.js prune --prefix=. --no-global && rimraf test/*/*/node_modul
es && make -j8 doc

require-inject@1.1.1 node_modules\require-inject

sprintf-js@1.0.2 node_modules\sprintf-js

marked@0.3.3 node_modules\marked

marked-man@0.1.4 node_modules\marked-man

nock@1.2.1 node_modules\nock
+-- propagate@0.3.1
+-- debug@1.0.4 (ms@0.6.2)
+-- lodash@2.4.1
+-- chai@1.10.0 (assertion-error@1.0.0, deep-eql@0.1.3)

npm-registry-mock@1.0.0 node_modules\npm-registry-mock
+-- util-extend@1.0.1
+-- hock@0.2.5 (deep-equal@0.2.1)

tap@0.7.1 node_modules\tap
+-- yamlish@0.0.6
+-- buffer-equal@0.0.1
+-- deep-equal@1.0.0
+-- glob@4.5.3
+-- difflet@0.2.6 (deep-is@0.1.3, charm@0.1.2, traverse@0.6.6)
+-- runforcover@0.0.2 (bunker@0.1.2)

npm-registry-couchapp@2.6.7 node_modules\npm-registry-couchapp
+-- json@9.0.3
+-- couchapp@0.11.0 (watch@0.8.0, coffee-script@1.9.1, url@0.10.3, connect@3.3.5
, nano@6.1.2, http-proxy@0.8.7)

D:\iojs\deps\npm>git add -A .

D:\iojs\deps\npm>git commit -m "deps: upgrade npm to 2.7.4"
[upgrade-npm 2e9a27c] deps: upgrade npm to 2.7.4
 632 files changed, 21000 insertions(+), 32646 deletions(-)
 delete mode 100644 deps/npm/.eslintrc
 create mode 100644 deps/npm/.gitignore
   << snip >>

D:\iojs\deps\npm>git checkout v1.x
Switched to branch 'v1.x'
Your branch is up-to-date with 'origin/v1.x'.

D:\iojs\deps\npm>git merge upgrade-npm
Removing deps/npm/test/tap/outdated-git/package.json
Removing deps/npm/test/tap/outdated-git/README.md
Removing deps/npm/test/fixtures/config/userconfig-with-gc
   << snip >>  
Merge made by the 'recursive' strategy.
 deps/npm/.eslintrc                                 |   20 -
 deps/npm/.gitignore                                |   30 +
 deps/npm/.mailmap                                  |    5 +
   << snip >>

D:\iojs\deps\npm>git log -n5 --oneline -- .
96c8f57 Merge branch 'upgrade-npm' into v1.x
2e9a27c deps: upgrade npm to 2.7.4
3d46fef win,node-gyp: allow node.exe/iojs.exe to be renamed
269e46b deps: make node-gyp work with io.js
b542fb9 deps: upgrade npm to 2.7.3
```

othiym23 commented:
@piscisaureus that's what I do, only with a little more rebasing along the way. If we have a stack of patches to apply / cherry-pick, though, a quilt is going to be less error-prone (for me, at least).

piscisaureus commented:
LINK : fatal error LNK1194: cannot delay-load 'iojs.exe' due to import of data
symbol '"__declspec(dllimport) const v8::OutputStream::`vftable'" (__imp_??_7Ou
tputStream@v8@@6B@)'; link without /DELAYLOAD:iojs.exe [D:\temp\node_modules\st
rong-pm\node_modules\strong-supervisor\node_modules\heapdump\build\heapdump.vcx
proj]

piscisaureus commented:
@bnoordhuis, @rvagg and @iojs/platform-windows 
See the above error message. @kkoopa kind of warned for this, but I couldn't reproduce it.
This is the only module I've gotten this error with so far, I thought somehow this had been fixed in newer versions of MSVS.
Anyhow, I still think we should keep the delay load hook because it fixes so much.
But should we make it opt-in or opt-out ?

bnoordhuis commented:
Does that happen with everything that inherits from a class with virtual methods?  Because in that case addons that create e.g. external strings won't work either.

piscisaureus commented:
@bnoordhuis Not sure, it didn't happen when I tested an addon with a class that inherits from `ExternalStringResource`.

am11 commented:
(not sure how relevant it is)
According to this: https://msdn.microsoft.com/en-us/library/yx1x886y.aspx

> Imports of data cannot be supported. A \*workaround\* is to explicitly handle the data import yourself using **LoadLibrary** (or **GetModuleHandle** after you know the delay-load helper has loaded the DLL) and **GetProcAddress**.

It seems like this guy has tried that workaround: http://www.cultdeadcow.com/tools/pewrap.html.
From: https://github.com/iojs/io.js/issues/751#issuecomment-82647866.

bnoordhuis commented:
Some quick googling suggests that disabling optimizations can fix the issue?  Not sure what to make of that (except 'compiler bug'.)

piscisaureus commented:
PR to make delay loading opt-in: https://github.com/iojs/io.js/pull/1266

zcbenz commented:
:sparkles: This change is extremely helpful for atom-shell users to rename the exe name! /cc @paulcbetts

kkoopa commented:
It's not a compiler bug, nor is it fixable. Data cannot be delayloaded because there is no way of intercepting its access. Exporting classes with virtual methods requires exporting a pointer to the vftable, this constitutes data and thus classes with virtual members cannot be exported.

Delayload works fine as a band-aid for most addons, but the shared library solution should always work. Being able to build io.js as a shared library is nice for many other reasons as well, so its priority should not be lowered.
